keps/sig-api-machinery/5647-stale-controller-handling/README.md

```diff
@@ -0,0 +1,815 @@
+version](https://github.com/kubernetes/enhancements/pull/5505) we now have the
+ability to have the idea of the "age" of a resource and run operations with that
+in mind.
+
+### Goals
+
+The goals of this KEP is to document the stale controller issue and propose
+solutions to help detect and mitigate it. We will propose the ideas of enforcing
+stricter cache semantics at certain points in the reconcile cycle of a
+controller.
+
+### Non-Goals
+
+This is not intended to enforce consistency guarantees on all controllers. The
+changes described here will be "opt-in" for every controller and will require
+some degree of changes to the controller logic. It is not the goal to require
+controllers to all view a consistent state of the world per reconcile.
+
+## Proposal
+
+We propose the ability for controllers to be able to view whether their own
+writes have made it into the apiserver. This will allow for the controller to
+pause their reconcile loop until it knows whether or not the previous write has
+been propogated to its cache. We will do this by adding plumbing to the
+underlying cached informer and adding the ability for controllers to know the
+resource version of objects they subscribe to. 
+
+Once this is added, we will onboard certain controllers to be able to use the
+newly exposed resource versions and pause their reconcile loops until certain
+objects they write to are updated in their cache.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I am a K8s cluster administrator. I want to be sure that my controllers do not
+write too frequently. I enable this feature so that I know my controllers only
+write when they are up to date.
+
+#### Story 2
+
+I am a controller author. I want certain objects to be ensured to be in my cache
+after I write on a previous reconcile so I can be assured that my reads are up
+to date. I use the newly provided  frameworks to ensure those objects are up to
+date, mitigating the risks that I am operating on stale data. 
+
+### Risks and Mitigations
+
+There is the risk of pausing the reconcile loop, if this is not correct and we
+don't unpause properly then that would lead to scenarios where a controller may
+stop reconciling an object entirely. We will feature gate this and add a set of
+rigorous tests prior to Beta/GA to ensure that the feature is well tested and
+not missing edge cases. Any time we try and optimize the reconcile loop there
+are issues like this that may arise.
+
+## Design Details
+
+Our proposal consists of two parts, one is the change that will enable the
+controllers to be fully informed on the current resource versions of objects in
+their cache. Second is the use of that new ability to actually ensure the read
+after write guarantees on objects we care about in the controller itself.
+
+### Informer and Cache Update
+
+We will update the `ResourceEventHandlerFuncs` in
+`staging/src/k8s.io/client-go/tools/cache/controller.go` to have one more function.
```

Comment:
nit: Adding functions to existing interfaces is breaking (since existing implementations developed out of tree won't have the new function). This is more of an implementation concern, but I'm curious if we should *change* the existing interface, or *add* a new interface like:

```
type BookmarkedResourceEventHandlerFuncs struct {
  ResourceEventHandlerFuncs
  BookmarkFunc func(resourceVersion string)
}
```

?